### ISSUE TITLE
Question: Rebuilding Font Atlas makes window transparent or font unreadable

### ISSUE THREAD

**Issue by tobiazsh** [SpaiR/imgui-java#299]
### Version

1.87.5

### What happened?

The window either goes black (unfortunately no screenshot) and the fonts disappear, or window goes transparent with only the outlines visible (screenshot attached) or the fonts are unreadable (screenshot attached) all when adding a new font to the font atlas and then rebuilding it. I don't know if this is an issue on my side or an issue with ImGui, but I suspect the latter.

**Unreadable font:**
![grafik](https://github.com/user-attachments/assets/8286c676-d4fc-4141-b5ae-d38a795a552c)
**Transparent window:**
![grafik](https://github.com/user-attachments/assets/83b6d551-7611-41b8-a1be-fad7e0f45f0d)

### Reproduction

When rebuilding the font, it happens.

**NOTE:** My code my seem weird but this is because I am using ImGui inside a Minecraft Context and thus I unfortunately have some limitations on what I can do. I cannot use Thread.sleep(); for example.

Here's what I've tried:
<br>
**First, I load the standard fonts on initialization:**
```java
    public static void create(final long handle) {
        ImGui.createContext();
        ImPlot.createContext();

        final ImGuiIO io = ImGui.getIO();
        io.setIniFilename(MyWorldTrafficAddition.MOD_ID + ".ini");
        io.setFontGlobalScale(1F);

        fontAtlas = io.getFonts();

        // Load fonts
        final ImFontGlyphRangesBuilder rangesBuilder = new ImFontGlyphRangesBuilder();
        rangesBuilder.addRanges(io.getFonts().getGlyphRangesDefault());
        rangesBuilder.addRanges(io.getFonts().getGlyphRangesCyrillic());
        rangesBuilder.addRanges(io.getFonts().getGlyphRangesJapanese());

        final ImFontConfig fontConfig = new ImFontConfig();

        final short[] glyphRanges = rangesBuilder.buildRanges();
        DejaVuSans = fontAtlas.addFontFromMemoryTTF(loadFromResource("/assets/myworld_traffic_addition/font/dejavu_sans.ttf"), 20, fontConfig, glyphRanges);
        DejaVuSansBold = fontAtlas.addFontFromMemoryTTF(loadFromResource("/assets/myworld_traffic_addition/font/dejavu_sans_bold.ttf"), 20, fontConfig, glyphRanges);
        DejaVuSansBoldBig = fontAtlas.addFontFromMemoryTTF(loadFromResource("/assets/myworld_traffic_addition/font/dejavu_sans_bold.ttf"), 40, fontConfig, glyphRanges);

        fontAtlas.build();

        fontConfig.destroy();

        io.setConfigFlags(ImGuiConfigFlags.DockingEnable);

        imGuiImplGlfw.init(handle, true);
        imGuiImplGl3.init();
    }
```
<br>

**Then, I render:**
```java
    public static void draw(final RenderInterface renderInterface) {
        processFontRequests();
        rebuildFontAtlasIfNeeded();

        imGuiImplGl3.newFrame();
        imGuiImplGlfw.newFrame();
        ImGui.newFrame();

        renderInterface.render(ImGui.getIO());

        ImGui.render();
        imGuiImplGl3.renderDrawData(ImGui.getDrawData());

        if (ImGui.getIO().hasConfigFlags(ImGuiConfigFlags.ViewportsEnable)) {
            final long pointer = GLFW.glfwGetCurrentContext();
            ImGui.updatePlatformWindows();
            ImGui.renderPlatformWindowsDefault();

            GLFW.glfwMakeContextCurrent(pointer);
        }
    }
```
<br>

**Then, I load the fonts during runtime:**
```java
    public static void processFontRequests() {
        while(!fontRequestQueue.isEmpty()) {
            FontRequest request = fontRequestQueue.poll();

            ImGuiFont existingFont = getExistingFont(request.path, request.fontSize);

            if (existingFont != null){
                request.future.complete(existingFont);
                continue;
            }

            ImGuiFont font = new ImGuiFont(request.path, ImGuiImpl.registerFont(request.path, request.fontSize));
            registeredFonts.add(font);
            request.future.complete(font);
        }
    }
```
Notes on that:
\- This happens with a future, because it's the only thing I found that Minecraft accepts.
\- This method is executed at the beginning of the `draw();` method, because ImGui
\- You can request a font registration and then it's stored in a List to register it next frame
\- Every font has their own class "`ImGuiFont`", but the font atlas is the same

<br>

**How I rebuild the font atlas:**
```java
    public static void rebuildFontAtlasIfNeeded() {
        if (fontsNeedRebuild) {
            // Rebuild the font atlas
            ImGui.getIO().getFonts().build();

            // Get font texture data
            ImInt width = new ImInt();
            ImInt height = new ImInt();
            ImInt bytesPerPixel = new ImInt();
            ByteBuffer textureData = ImGui.getIO().getFonts().getTexDataAsRGBA32(width, height, bytesPerPixel);

            // Upload the texture data to the GPU
            int texId = uploadTextureToGPU(textureData, width.get(), height.get());

            // Set the new texture ID for ImGui
            ImGui.getIO().getFonts().setTexID(texId);

            // Notify the renderer (OpenGL) to create the font texture
            imGuiImplGl3.createFontsTexture();

            // Mark that the fonts have been rebuilt
            fontsNeedRebuild = false;
        }
    }
```

<br>

**Uploading textures to GPU (read somewhere that rebuilding the font atlas makes the old textures invalid and thus I do this):**
```java
    private static int uploadTextureToGPU(ByteBuffer textureData, int width, int height) {
        // Generate and bind the texture
        int[] texId = new int[1];
        GL33.glGenTextures(texId);
        GL33.glBindTexture(GL33.GL_TEXTURE_2D, texId[0]);

        // Upload the texture data
        GL33.glTexImage2D(
                GL33.GL_TEXTURE_2D,
                0,
                GL33.GL_RGBA,
                width,
                height,
                0,
                GL33.GL_RGBA,
                GL33.GL_UNSIGNED_BYTE,
                textureData
        );

        // Set texture parameters (important for ImGui text rendering)
        GL33.glTexParameteri(GL33.GL_TEXTURE_2D, GL33.GL_TEXTURE_MIN_FILTER, GL33.GL_LINEAR);
        GL33.glTexParameteri(GL33.GL_TEXTURE_2D, GL33.GL_TEXTURE_MAG_FILTER, GL33.GL_LINEAR);
        GL33.glTexParameteri(GL33.GL_TEXTURE_2D, GL33.GL_TEXTURE_WRAP_S, GL33.GL_CLAMP_TO_EDGE);
        GL33.glTexParameteri(GL33.GL_TEXTURE_2D, GL33.GL_TEXTURE_WRAP_T, GL33.GL_CLAMP_TO_EDGE);

        // Unbind the texture (optional but good practice)
        GL33.glBindTexture(GL33.GL_TEXTURE_2D, 0);

        return texId[0]; // Return the OpenGL texture ID
    }
```

<br><br>
This is how I read the fonts from memory (they are stored inside a .jar in the resources):
```java
    private static byte[] loadFromResource(String name) {
        try {
            return Files.readAllBytes(Paths.get(MyWorldTrafficAdditionClient.class.getResource(name).toURI()));
        } catch (URISyntaxException | IOException e) {
            throw new RuntimeException(e);
        }
    }
```

**Comment by SpaiR** [comment]
The issue you described is solely a problem with your code. At a minimum, before calling `ImGui.getIO().getFonts().build();`, you need to call `ImGui.getIO().getFonts().clear();`, which is missing in your case. Same for the `createFontsTexture()`, u need to call `destroyFontsTexture()`. I'll leave the issue (question) open in case someone else wants to contribute additional feedback, but most likely, you'll need to resolve the issue yourself. I recommend reviewing the relevant materials on Dear ImGui.

**Comment by tobiazsh** [comment]
> The issue you described is solely a problem with your code. At a minimum, before calling `ImGui.getIO().getFonts().build();`, you need to call `ImGui.getIO().getFonts().clear();`, which is missing in your case. Same for the `createFontsTexture()`, u need to call `destroyFontsTexture()`. I'll leave the issue (question) open in case someone else wants to contribute additional feedback, but most likely, you'll need to resolve the issue yourself. I recommend reviewing the relevant materials on Dear ImGui.

Thank you so much for commenting on this! I have been trying to get this working for such a long time now and I've only used code that I've found online about the font atlas.

Thanks for clarifying!

**Comment by tobiazsh** [comment]
@SpaiR Your help has helped me out so much! Thank you!

I was able to solve this problem by:

1. Clearing the font atlas (`fontAtlas.clear();`)
2. Re-adding the old fonts (that are preferably stored somewhere; e.g. in a List<>) and then re-defining their ImFont instance.
3. Building the font atlas (`fontAtlas.build();`)
4. Uploading the font textures to GPU (MAY not be necessary, but I did that anyways)
5. Telling the graphics lib to create the font textures (MAY not be necessary; In my case: `ImGuiImplGL3.createFontTextures();`
